src/test/java/com/google/edith/LoginServletTest.java

```diff
@@ -0,0 +1,105 @@
+
+import com.google.edith.servlets.LoginServlet;
+
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.atLeast;
+import org.mockito.MockitoAnnotations;
+import org.mockito.ArgumentCaptor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.PrintWriter;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public final class LoginServletTest {
+  
+  private Map<String, Object> myMap = new HashMap<String, Object>() {{
+        put("com.google.appengine.api.users.UserService.user_id_key", "12345");
+    }};
+
+  private final LocalServiceTestHelper loggedInTestHelper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+      .setEnvAttributes(myMap)
+      .setEnvIsLoggedIn(true)
+      .setEnvAuthDomain("gmail")
+      .setEnvIsAdmin(true)
+      .setEnvEmail("user@gmail.com");
+      
+  private final LocalServiceTestHelper loggedOutTestHelper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+      .setEnvIsLoggedIn(false);
+  
+  private final UserService userService = UserServiceFactory.getUserService();
+  private LoginServlet loginServlet;
+
+  @Mock
+  HttpServletRequest request;
+
+  @Mock
+  HttpServletResponse response;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    loginServlet = new LoginServlet();
+  }
+
+  @Test
+  public void testWhenLoggedIn() throws IOException, ServletException {
+    loggedInTestHelper.setUp();
+    assertTrue(userService.isUserLoggedIn());
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    loginServlet.doGet(request, response);
+    verify(response, atLeast(1)).getWriter();
+    loggedInTestHelper.tearDown();
+  }
+
+  @Test
+  public void testWhenLoggedOut() throws IOException, ServletException {
+    loggedOutTestHelper.setUp();
+    assertFalse(userService.isUserLoggedIn());
+    ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+    loginServlet.doGet(request, response);
+    verify(response).sendRedirect(captor.capture());
+    assertEquals(userService.createLoginURL("/"), captor.getValue());
+    loggedOutTestHelper.tearDown();
```

Comment:
could teardown and setup go in before and after tags as well?

Reply:
Two setups have to be done to mimic user being logged in and logged out. That is why I could not put them in before and after